websockets/protocol/caret.go

```diff
@@ -0,0 +1,81 @@
+package protocol
+
+type Caret struct {
+	Start int `json:"caret_start"`
+	End   int `json:"caret_end"`
+}
+
+func processAdd(
+	receiverCaret Caret,
+	senderCaret Caret,
+	caretStartDelta int,
+	caretEndDelta int,
+	docDelta int,
+) Caret {
+	if senderCaret.End < receiverCaret.End && caretEndDelta != 0 {
+		receiverCaret.End += docDelta
+		if senderCaret.End <= receiverCaret.Start {
+			receiverCaret.Start += docDelta
+		}
+	}
+	return receiverCaret
+}
+
+func ShiftCaret(
+	receiverCaret Caret,
+	senderCaret Caret,
+	caretStartDelta int,
+	caretEndDelta int,
+	docDelta int,
```

Comment:
Oh actually, it might be cleaner if the parameter was just a `Delta` object.

Reply:
Yeah you're right, I just wanted to avoid dealing with pointer logic.

Reply:
Made some other changes to to these functions. Made them `Caret` methods and just made `processAdd` accept a single `delta` integer.